File: generate-svgs.swift
#!/usr/bin/env swift
// Generates SVG files from LucideIcons.swift icon data into Resources/icons/
// Run: swift generate-svgs.swift

import Foundation

let sourcePath = "Sources/MenuBarFolders/Services/LucideIcons.swift"
let outputDir = "Resources/icons"

guard let source = try? String(contentsOfFile: sourcePath, encoding: .utf8) else {
    print("Could not read \(sourcePath)")
    exit(1)
}

try? FileManager.default.createDirectory(atPath: outputDir, withIntermediateDirectories: true)

// Parse icon entries from the Swift source
// Format: "icon-name": [#"<element/>"#, #"<element/>"#],
let pattern = #"\"([a-z0-9-]+)\"\s*:\s*\[((?:#\"[^\"]*\"#(?:,\s*)?)+)\]"#
guard let regex = try? NSRegularExpression(pattern: pattern) else {
    print("Regex failed"); exit(1)
}

let range = NSRange(source.startIndex..., in: source)
let matches = regex.matches(in: source, range: range)

var count = 0
for match in matches {
    guard let nameRange = Range(match.range(at: 1), in: source),
          let elemRange = Range(match.range(at: 2), in: source) else { continue }

    let name = String(source[nameRange])
    let rawElements = String(source[elemRange])

    // Extract individual element strings from #"..."#
    let elemPattern = #"#\"([^\"]+)\"#"#
    guard let elemRegex = try? NSRegularExpression(pattern: elemPattern) else { continue }
    let elemNSRange = NSRange(rawElements.startIndex..., in: rawElements)
    let elemMatches = elemRegex.matches(in: rawElements, range: elemNSRange)

    var elements: [String] = []
    for em in elemMatches {
        guard let r = Range(em.range(at: 1), in: rawElements) else { continue }
        elements.append(String(rawElements[r]))
    }

    let svg = """
    <svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round">
      \(elements.joined(separator: "\n  "))
    </svg>
    """

    let path = "\(outputDir)/\(name).svg"
    try? svg.write(toFile: path, atomically: true, encoding: .utf8)
    count += 1
}

print("Generated \(count) SVG files in \(outputDir)/")

File: generate-svgs.py
#!/usr/bin/env python3
"""Generate SVG files from LucideIcons.swift icon data into Resources/icons/"""

import re
import os

SOURCE = "Sources/MenuBarFolders/Services/LucideIcons.swift"
OUTPUT = "Resources/icons"

os.makedirs(OUTPUT, exist_ok=True)

with open(SOURCE) as f:
    source = f.read()

# Find all icon entries: "name": [#"<elem/>"#, ...]
# Parse the icon dictionary block
icon_pattern = re.compile(r'"([a-z0-9-]+)"\s*:\s*\[([^\]]+)\]')
elem_pattern = re.compile(r'#"(.+?)"#')

count = 0
for match in icon_pattern.finditer(source):
    name = match.group(1)
    raw = match.group(2)
    elements = elem_pattern.findall(raw)
    if not elements:
        continue

    indent = "  "
    elems_str = f"\n{indent}".join(elements)
    svg = f'''<svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round">
  {elems_str}
</svg>
'''

    with open(os.path.join(OUTPUT, f"{name}.svg"), "w") as f:
        f.write(svg)
    count += 1

print(f"Generated {count} SVG files in {OUTPUT}/")
